### DISCUSSION TITLE
Problem when downgrading from eclipse-zenoh 0.5.0-beta.9 to beta.8

### DISCUSSION THREAD

**Discussion by paloma-jim** [eclipse-zenoh/roadmap#62] [Q&A] [answered]
Hello,

I am working with Zenoh with a publisher injecting frames in the channel and a worker reading the frames. However, I realized that by mistake the publisher has eclipse-zenoh==0.5.0b9 and the worker has eclipse-zenoh==0.5.0b8 and this causes problems.

I am interested in continue working with version 0.5.0b8 because of other dependencies, however, when I run `pip3 install eclipse-zenoh==0.5.0b8`, I get the following error:

Do you know what is wrong? How can I downgrade the eclipse-zenoh version of the publisher?

```
ERROR: Could not find a version that satisfies the requirement eclipse-zenoh==0.5.0b8 (from versions: 0.4.2.dev3, 0.4.2.dev5, 0.4.2rc1, 0.5.0b4, 0.5.0b5, 0.5.0b7, 0.5.0b8, 0.5.0b9)
ERROR: No matching distribution found for eclipse-zenoh==0.5.0b8
```
I am using ubuntu 18.04.

Thank you

**Comment by JEnoch** [comment, marked as answer]
Hi @paloma-jim,

The version 0.5.0-beta.8 for the Zenoh Python API was named **`0.5.0-b8`** (see https://pypi.org/project/eclipse-zenoh/0.5.0-b8/#history).
So you can install it running: `pip install eclipse-zenoh==0.5.0-b8`

**Reply by paloma-jim** [reply]
Hi @JEnoch 

Thank you for your answer. The problem still continues when i try `pip3 install eclipse-zenoh==0.5.0-b8`.  This is the further description of the problem:
```
Collecting eclipse-zenoh==0.5.0-b8
  Using cached eclipse_zenoh-0.5.0_b8.tar.gz (74 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3.6 /usr/local/lib/python3.6/dist-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /tmp/tmpkhmgrqy_
       cwd: /tmp/pip-install-i9kplwq2/eclipse-zenoh_6a7c1e32489249bcbdc0a9a601dd8994
  Complete output (20 lines):
  \xe2\x9a\xa0\xef\xb8\x8f  Warning: Please use maturin in pyproject.toml with a version constraint, e.g. `requires = ["maturin>=0.12,<0.13"]`. This will become an error.
  \xf0\x9f\x92\xa5 maturin failed
    Caused by: Cargo metadata failed. Does your crate compile with `cargo build`?
    Caused by: `cargo metadata` exited with an error:     Updating crates.io index
   Downloading crates ...
  error: failed to download `indexmap v1.9.1`

  Caused by:
    unable to get packages from source

  Caused by:
    failed to parse manifest at `/root/.cargo/registry/src/github.com-1ecc6299db9ec823/indexmap-1.9.1/Cargo.toml`

  Caused by:
    feature `edition2021` is required

    consider adding `cargo-features = ["edition2021"]` to the manifest
  Error running maturin: Command '['maturin', 'pep517', 'write-dist-info', '--metadata-directory', '/tmp/pip-modern-metadata-8vv4fu8f', '--interpreter', '/usr/bin/python3.6']' returned non-zero exit status 1.
  Checking for Rust toolchain....
  Running `maturin pep517 write-dist-info --metadata-directory /tmp/pip-modern-metadata-8vv4fu8f --interpreter /usr/bin/python3.6`
  ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/5f/55/5067f451752abf3ed060fabd6f53a071bcae8b080430574575b63d9d3171/eclipse_zenoh-0.5.0_b8.tar.gz#sha256=922f7de5ea85d4fff171e4dcedb38d9ff20471398a2f44ac1c870db65d0eff2a (from https://pypi.org/simple/eclipse-zenoh/). Command errored out with exit status 1: /usr/bin/python3.6 /usr/local/lib/python3.6/dist-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /tmp/tmpkhmgrqy_ Check the logs for full command output.
ERROR: Could not find a version that satisfies the requirement eclipse-zenoh==0.5.0-b8 (from versions: 0.4.2.dev3, 0.4.2.dev5, 0.4.2rc1, 0.5.0b4, 0.5.0b5, 0.5.0b7, 0.5.0b8, 0.5.0b9)
ERROR: No matching distribution found for eclipse-zenoh==0.5.0-b8
```
Do you know what could be wrong?
Thank you

**Reply by JEnoch** [reply]
When releasing zenoh-python 0.5.0-b8, we didn't commit the Cargo.lock file generated by Rust which fixes all the versions of dependencies used by zenoh-python. Thus, this file isn't present either in the source package (eclipse_zenoh-0.5.0_b8.tar.gz).

As a consequence, Rust uses more recent versions of those dependencies than the ones used at the time of 0.5.0-b8 release.
And those recent dependencies are relying on a more recent Rust edition (2021) than the Rust toolchain that was configured for 0.5.0-b8.

I managed to build again this 0.5.0-b8 from sources with recent Rust (nightly) toolchain with the following commands that you should try:
```bash
# make sure Maturin 0.10.0 is installed
pip3 uninstall maturin
pip3 install maturin==0.10.0
~/.local/bin/maturin --version

# get 0.5.0-b8 sources package and unzip it
wget https://files.pythonhosted.org/packages/5f/55/5067f451752abf3ed060fabd6f53a071bcae8b080430574575b63d9d3171/eclipse_zenoh-0.5.0_b8.tar.gz
mkdir zenoh-python-0.5.0-b8
cd zenoh-python-0.5.0-b8
tar xzf ../eclipse_zenoh-0.5.0_b8.tar.gz

# update rust-toolchain to use latest nightly version
echo "nightly" > rust-toolchain

# generate and update Cargo.lock with appropriate dependencies versions for 0.5.0-b8
cargo update -p zenoh --precise 0.5.0-beta.8
cargo update -p zenoh-util --precise 0.5.0-beta.8
cargo update -p nix@0.20.2 --precise 0.20.0
cargo update -p nix@0.22.3 --precise 0.20.0

# buid the wheel forcing target to gnu (maturin defaults to musl... not sure why)
~/.local/bin/maturin build --release --target `uname -m`-unknown-linux-gnu
# install the resulting wheel
pip3 install ./target/wheels/eclipse_zenoh-0.5.0_b8*.whl
```

Let me know if this works for you.

BTW, some built binary wheels for 0.5.0_b8 are also available (see https://pypi.org/project/eclipse-zenoh/0.5.0-b8/#files) but for a limited set of Linux platforms (x86_64, i686 and armv7).
I guess you're on a different platform, and that's why `pip3` selects the source distribution, right ?
You might not have the same issue on a x86_64 platform, where the binary wheel will be installed.

**Reply by paloma-jim** [reply]
Hi @JEnoch ,

It worked properly. Thank you very much for your help.

I am working with Linux x86_64. When proceeding as explained above, everything worked as expected :)
